File: test/test_supabase_connection.py
"""Tests for the Supabase PostgreSQL connection helpers.

The live connection check is intentionally skipped unless a Supabase database URL
is provided through the same environment variables used by the application.
"""
from __future__ import annotations

import os
import unittest

from sqlalchemy import text
from sqlalchemy.engine import URL

from db import connection
from utils.logging import get_logger


LOGGER = get_logger("db_test")


class SupabaseConnConfigTest(unittest.TestCase):
    """Validate Supabase database URL configuration behavior."""

    def setUp(self) -> None:
        self._orig_env = {
            name: os.environ.get(name)
            for name in connection.DATABASE_URL_ENV_NAMES
        }
        for name in self._orig_env:
            os.environ.pop(name, None)
        connection.get_engine.cache_clear()

    def tearDown(self) -> None:
        connection.close_db()
        for name, value in self._orig_env.items():
            if value is None:
                os.environ.pop(name, None)
            else:
                os.environ[name] = value

    def test_db_url_normalized(self) -> None:
        """Plain PostgreSQL URLs are normalized to SQLAlchemy's psycopg driver."""
        sample_url = URL.create(
            drivername="postgresql",
            username="postgres",
            password="secret",
            host="db.example.supabase.co",
            port=5432,
            database="postgres",
        )
        os.environ["DATABASE_URL"] = sample_url.render_as_string(hide_password=False)

        url = connection.get_database_url()

        self.assertEqual(
            url,
            sample_url.set(drivername="postgresql+psycopg").render_as_string(
                hide_password=False
            ),
        )

    def test_missing_db_url_raises(self) -> None:
        """A clear error is raised when no supported connection URL is set."""
        with self.assertRaisesRegex(RuntimeError, "DATABASE_URL"):
            connection.get_database_url()


@unittest.skipUnless(
    any(os.getenv(env_name) for env_name in connection.DATABASE_URL_ENV_NAMES)
    and any(os.getenv(env_name) for env_name in connection.SECRET_KEY_ENV_NAMES),
    "Set DATABASE_URL and SECRET_KEY to run the live Supabase connection test.",
)
class SupabaseConnLiveTest(unittest.TestCase):
    """Live Supabase connectivity test."""

    def tearDown(self) -> None:
        connection.close_db()

    def test_executes_select_one(self) -> None:
        """The configured Supabase database accepts a simple SQL round trip."""
        with connection.db_session() as session:
            val = session.execute(text("SELECT 1")).scalar_one()

        self.assertEqual(val, 1)
        LOGGER.info("Live database connection succeeded.")


if __name__ == "__main__":
    unittest.main()

File: test/test_supabase_connection.js
import assert from 'node:assert/strict';
import test from 'node:test';

import { closeDb, getConnection, getDatabaseUrl, redactedDatabaseUrl } from '../db/connection.js';

test('getDatabaseUrl returns configured database URL', () => {
  const original = process.env.DATABASE_URL;
  process.env.DATABASE_URL = 'postgresql://user:secret@db.example.com:5432/postgres';
  assert.equal(getDatabaseUrl(), 'postgresql://user:secret@db.example.com:5432/postgres');
  if (original === undefined) delete process.env.DATABASE_URL; else process.env.DATABASE_URL = original;
});

test('getDatabaseUrl raises when no URL is configured', () => {
  const original = process.env.DATABASE_URL;
  delete process.env.DATABASE_URL;
  assert.throws(() => getDatabaseUrl(), /DATABASE_URL/);
  if (original === undefined) delete process.env.DATABASE_URL; else process.env.DATABASE_URL = original;
});

test('redactedDatabaseUrl hides passwords', () => {
  assert.equal(redactedDatabaseUrl('postgresql://user:secret@db.example.com:5432/postgres'), 'postgresql://user:***@db.example.com:5432/postgres');
});

test('getConnection lazily returns connection metadata when enabled', async () => {
  const originalDatabaseUrl = process.env.DATABASE_URL;
  const originalSecretKey = process.env.SECRET_KEY;
  process.env.DATABASE_URL = 'postgresql://user:secret@db.example.com:5432/postgres';
  process.env.SECRET_KEY = 'secret';
  closeDb();
  const connection = await getConnection();
  assert.equal(connection.connected, false);
  assert.match(connection.redactedUrl, /\*\*\*/);
  closeDb();
  if (originalDatabaseUrl === undefined) delete process.env.DATABASE_URL; else process.env.DATABASE_URL = originalDatabaseUrl;
  if (originalSecretKey === undefined) delete process.env.SECRET_KEY; else process.env.SECRET_KEY = originalSecretKey;
});
